includes/Modules/Analytics/Advanced_Tracking/Measurement_Events/WooCommerce_Event_List.php

```diff
@@ -18,134 +22,346 @@
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param WC_Cart $cart The WooCommerce cart instance.
+	 * @throws \Exception Thrown when invalid keys or value type.
+	 */
+	private function create_wc_checkout_event( $cart ) {
+		$items = array();
+		foreach ( $cart->get_cart() as $cart_item ) {
+			$product         = $cart_item['data'];
+			$item            = array();
+			$category_string = '';
+			$category_ids    = $product->get_category_ids();
+			foreach ( $category_ids as $category_id ) {
+				$category_query = get_term_by( 'id', $category_id, 'product_cat' );
+				if ( 'WP_Term' == get_class( $category_query ) ) {
+					$category_string .= $category_query->name . '/';
+				}
+			}
+			$item['category'] = substr( $category_string, 0, strlen( $category_string ) - 1 );
+			$item['id']       = $product->get_sku();
+			$item['name']     = $product->get_name();
+			$item['price']    = $product->get_price();
+			$item['quantity'] = $cart_item['quantity'];
+			$items[]          = $item;
+		}
+
+		$checkout_meta                   = array();
+		$checkout_meta['event_category'] = 'ecommerce';
+		$checkout_meta['value']          = $cart->get_subtotal();
+		$checkout_meta['currency']       = get_woocommerce_currency();
+		$checkout_meta['items']          = $items;
+		$checkout_event                  = new Measurement_Event(
 			array(
 				'pluginName' => 'WooCommerce',
-				'category'   => 'ecommerce',
-				'action'     => 'purchase',
-				'selector'   => '.woocommerce-page form.woocommerce-checkout',
-				'on'         => 'submit',
+				'action'     => 'begin_checkout',
+				'on'         => 'DOMContentLoaded',
+				'metadata'   => $checkout_meta,
 			)
 		);
-		$this->add_event( $event );
+		$this->add_event( $checkout_event );
+	}
+
+	/**
+	 * Creates a purchase Measurement_Event object when an order is received.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param WC_Order $order The WooCommerce order.
+	 * @throws \Exception Thrown when invalid keys or value type.
+	 */
+	private function create_wc_purchase_event( $order ) {
+		$items       = array();
+		$order_items = $order->get_items( apply_filters( 'woocommerce_purchase_order_item_types', 'line_item' ) );
+		foreach ( $order_items as $item_id => $item ) {
+			$product         = $item->get_product();
+			$purchase_item   = array();
+			$category_string = '';
+			$category_ids    = $product->get_category_ids();
+			foreach ( $category_ids as $category_id ) {
+				$category_query = get_term_by( 'id', $category_id, 'product_cat' );
+				if ( 'WP_Term' == get_class( $category_query ) ) {
+					$category_string .= $category_query->name . '/';
+				}
+			}
+			$item['category']          = substr( $category_string, 0, strlen( $category_string ) - 1 );
+			$purchase_item['id']       = $product->get_sku();
+			$purchase_item['name']     = $product->get_name();
+			$purchase_item['price']    = $product->get_price();
+			$purchase_item['quantity'] = $item->get_quantity();
```

Comment:
Actually, let's do something else instead of introducing a `parse_category( $product )` method: Since all the products are parsed into an item for event tracking with exactly the same logic, let's introduce a private method `parse_product_into_item( \WC_Product $product )`, which returns an item associative array (with `category`, `id`, `name`, `price`, `quantity`). This gets rid of a lot of duplicate code here.

Reply:
Sounds good, but I think it would be a good idea to leave quantity out of the array since different contexts handle that key differently.